### PR TITLE
Fixes #22747. Support [crop] configuration for DarkNet

### LINKED ISSUE #22747: darknet crop layer missing implementation
### System Information

OpenCV python version: 4.6.0
Python version: 3.10.8
Operating System / Platform: Archlinux 6.0.2-arch1-1

### Detailed description

Can't create the Vgg-16 net from DarkNet

```Error: Unknown layer type: crop in function 'ReadDarknetFromCfgStream'```

### Steps to reproduce

```python
cv2.dnn.readNetFromDarknet('vgg-16.cfg', '') # model is not needed to reproduce the issue
```
[vgg-16.cfg](https://raw.githubusercontent.com/pjreddie/darknet/master/cfg/vgg-16.cfg)

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by Dhanwanth1803**
Request for comments. This is my first PR. 

**Merge with extra**: https://github.com/opencv/opencv_extra/pull/1112

resolves https://github.com/opencv/opencv/issues/22747

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
opencv_extra=feat-crop1
```

### CODE REVIEW

**Comment by opencv-alalek** [comment]
It makes sense to add test to avoid regressions in the future
- see test code for `Test_Darknet_layers`
- related test data is here: https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet

**Comment by Dhanwanth1803** [comment]
> It makes sense to add test to avoid regressions in the future
> - see test code for `Test_Darknet_layers`
> - related test data is here: https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet

Yes sure @opencv-alalek. I will add the tests very soon. Can you please assign this issue to me.

**Comment by Dhanwanth1803** [comment]
@dkurt @opencv-alalek Hi guys, I am trying to add the test layers. Its actually a lot of test layers and I am finding it hard to understand the parameters and add it accordingly. Can you guys help me with any resources or tell me how can I go about this problem. It would be a great help. thanks :)

**Comment by dkurt** [comment]
@Dhanwanth1803, you can create a single layer network similar to models from https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet. Then, use a script [generate_darknet_models.py](https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/generate_darknet_models.py) to create test data files. As Crop layer has no weights, it's enough to write only a `.cfg` file. Take a look at average pooling as a reference: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/avgpool_softmax.cfg

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803, you can create a single layer network similar to models from https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet. Then, use a script [generate_darknet_models.py](https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/generate_darknet_models.py) to create test data files. As Crop layer has no weights, it's enough to write only a `.cfg` file. Take a look at average pooling as a reference: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/avgpool_softmax.cfg
> 
> After that, add all the generated files to a new branch to opencv_extra with the same name: `[feat-crop](https://github.com/Dhanwanth1803/opencv/tree/feat-crop)` and open a PR.

Thank you @dkurt. I will be working on it.

**Comment by Dhanwanth1803** [comment]
@dkurt Like you said, i tried to generate test data files using [generate_darknet_models.py](https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/generate_darknet_models.py) . But it gave me the following error. 
AttributeError: module 'darknet' has no attribute 'load_net'
So I checked for load_net function in the darknet repo as well. Turns out there is no such function there. There is a load_network function but i tried using it already. Its not helping.

**Comment by dkurt** [comment]
@Dhanwanth1803, check https://github.com/pjreddie/darknet/blob/f6afaabcdf85f77e7aff2ec55c020c0e297c77f9/python/darknet.py#L85

**Comment by Dhanwanth1803** [comment]
@dkurt I have installed darknet and set the path using `export` and also i have given read and write permissions to the config files in `testdata/dnn/darknet`. when the run the `generate_darknet_models.py` I am getting the following error message.

`Couldn't open file: shortcut.cfg` 

This is happening for every function call trying to read a `.cfg` file.

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803, Run the script from `opencv_extra/testdata/dnn/darknet` folder

Yeah. I did run it from `opencv_extra/testdata/dnn/darknet`.

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803, you can create a single layer network similar to models from https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet. Then, use a script [generate_darknet_models.py](https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/generate_darknet_models.py) to create test data files. As Crop layer has no weights, it's enough to write only a `.cfg` file. Take a look at average pooling as a reference: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/avgpool_softmax.cfg
> 
> After that, add all the generated files to a new branch to opencv_extra with the same name: `[feat-crop](https://github.com/Dhanwanth1803/opencv/tree/feat-crop)` and open a PR.

@dkurt I have put up PR in `opencv_extra`. Please do check it out and let me know if any changes are required. Thank you.

**Comment by Dhanwanth1803** [comment]
> > @Dhanwanth1803, you can create a single layer network similar to models from https://github.com/opencv/opencv_extra/tree/4.x/testdata/dnn/darknet. Then, use a script [generate_darknet_models.py](https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/generate_darknet_models.py) to create test data files. As Crop layer has no weights, it's enough to write only a `.cfg` file. Take a look at average pooling as a reference: https://github.com/opencv/opencv_extra/blob/4.x/testdata/dnn/darknet/avgpool_softmax.cfg
> > After that, add all the generated files to a new branch to opencv_extra with the same name: `[feat-crop](https://github.com/Dhanwanth1803/opencv/tree/feat-crop)` and open a PR.
> 
> @dkurt I have put up PR in `opencv_extra`. Please do check it out and let me know if any changes are required. Thank you.

I have put up a PR again @dkurt . Please check it out.

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 1]
```diff
@@ -178,6 +178,26 @@ namespace cv {
                     fused_layer_names.push_back(last_layer);
                 }
 
+                void setCrop(int crop_height, int crop_width, int flip, int exposure, int saturation, int angle)
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:181]: Unused arguments. How layer can work without them?

**Comment by dkurt** [comment]
@Dhanwanth1803, add a test to https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_darknet_importer.cpp

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803, add a test to https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_darknet_importer.cpp

Okay. Will do.

**Comment by asmorkalov** [comment]
@Dhanwanth1803 Friendly reminder.

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803 Friendly reminder.

@asmorkalov I have exams currently in my school right now. I continue on 8th November. Is that fine?

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803 Friendly reminder.

have exams currently in my school right now. I continue on 8th November. Is that fine?

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803, add a test to https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_darknet_importer.cpp

@dkurt Should I add a test to 

> @Dhanwanth1803, add a test to https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_darknet_importer.cpp

@dkurt I have made changes for addition of crop layer in `https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/darknet/darknet_io.cpp` and test in `https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_darknet_importer.cpp`. Should I be making changes to any new file in addition to these two?

**Comment by asmorkalov** [comment]
@Dhanwanth1803 the PR still includes changes for single file. Looks like you forgot to add test changes to the commit.

**Comment by Dhanwanth1803** [comment]
> @Dhanwanth1803 the PR still includes changes for single file. Looks like you forgot to add test changes to the commit.

@asmorkalov @dkurt Yes Sir, I still didnt add the changes yet. I have a doubt. The branch in which I made changes and put up this PR. I think I lost that branch because I changed my pc. How should I add new changes in a diff branch to this PR?

**Comment by dkurt** [comment]
Please remove an old branch https://github.com/Dhanwanth1803/opencv_extra/tree/feat-crop

**Comment by Dhanwanth1803** [comment]
> Please remove an old branch https://github.com/Dhanwanth1803/opencv_extra/tree/feat-crop

Yes Sir deleted just now/

**Comment by dkurt** [comment]
@Dhanwanth1803, please make test pass locally. Build OpenCV with `ts` module and run `opencv_test_dnn --gtest_filter=Test_Darknet_layers.crop/0`

For now problem is not solved:
```
[ RUN      ] Test_Darknet_layers.crop/0, where GetParam() = OCV/CPU
unknown file: error: C++ exception with description "OpenCV(4.8.0-dev) c:\build\precommit_windows64\4.x\opencv\modules\dnn\src\net_impl.hpp:108: error: (-2:Unspecified error) Can't create layer "crop_0" of type "CropLayer" in function 'cv::dnn::dnn4_v20230620::Net::Impl::getLayerInstance'
" thrown in the test body.
[  FAILED  ] Test_Darknet_layers.crop/0, where GetParam() = OCV/CPU (1 ms)
```

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 2]
```diff
@@ -308,6 +308,26 @@ namespace cv {
...
                     fused_layer_names.push_back(last_layer);
                 }
+                
+                void setCrop(int crop_height, int crop_width)
+                {
+                    cv::dnn::LayerParams crop_param;
+                    crop_param.set<int>("crop_height", crop_height);
+                    crop_param.set<int>("crop_width", crop_width);
+                    crop_param.name = "CropLayer-name";
+                    crop_param.type = "CropLayer";
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:318]: There is no "CropLayer" in OpenCV - just "Crop"

File: modules/dnn/src/darknet/darknet_io.cpp [hunk 3]
```diff
@@ -760,6 +780,16 @@ namespace cv {
                         tensor_shape[1] = 1;
                         tensor_shape[2] = 1;
                     }
+                    else if (layer_type == "crop")
+                    {
+                        int crop_height = getParam<int>(layer_params, "crop_height", 3);
+                        int crop_width = getParam<int>(layer_params, "crop_width", 2);
```
Comment (dkurt) [inline modules/dnn/src/darknet/darknet_io.cpp:786]: Let's avoid using default values here - the arguments are required.

Comment (Dhanwanth1803) [reply modules/dnn/src/darknet/darknet_io.cpp:786]: @dkurt Yes sir will correct them.

File: modules/dnn/src/darknet/darknet_io.cpp (hunk 2 above)
Comment (dkurt) [reply modules/dnn/src/darknet/darknet_io.cpp:318]: Moreover, there is no `crop_height` and `crop_width` attributes for Crop. You need to utilize Slice with it's `begin`, `end` (https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/layers/slice_layer.cpp):

```cpp
std::vector<int> begin = {0, 0, 0, 0};
std::vector<int> sizes = {-1, -1, crop_height, crop_width};
crop_param.set("begin", DictValue::arrayInt(&begin[0], begin.size()));
crop_param.set("size", DictValue::arrayInt(&sizes[0], sizes.size()));
crop_param.type = "Slice";
```

Comment (Dhanwanth1803) [reply modules/dnn/src/darknet/darknet_io.cpp:318]: if `crop_height` and `crop_width` are not attributes for crop. Why did we generate test data using those parameters in `opencv_extra`

File: modules/dnn/src/darknet/darknet_io.cpp (hunk 3 above)
Comment (Dhanwanth1803) [reply modules/dnn/src/darknet/darknet_io.cpp:786]: What can I use instead of default values? The function demands initial values.

Comment (dkurt) [reply modules/dnn/src/darknet/darknet_io.cpp:786]: Use zero values and check that after retrieving they are not zero (means values are presented)

**Comment by asmorkalov** [comment]
@Dhanwanth1803  Friendly reminder

**Comment by asmorkalov** [comment]
@dkurt 
```
[ RUN      ] Test_Darknet_layers.crop/0, where GetParam() = OCV/CPU
unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /home/ci/opencv/modules/ts/src/ts.cpp:1071: error: (-2:Unspecified error) OpenCV tests: Can't find required data file: dnn/darknet/crop_in.npy in function 'findData'
" thrown in the test body.
[  FAILED  ] Test_Darknet_layers.crop/0, where GetParam() = OCV/CPU (0 ms)
```

**Comment by dkurt** [comment]
@asmorkalov, GitHub Actions CI ignores **Merge with extra** which points to a different branch name. Buildbot CI passes.

**Comment by asmorkalov** [comment]
@dkurt Extra CI options are handled by buildbot only. Please ignore Github actions for now.

**Review by dkurt — APPROVED** [review]

**Comment by dkurt** [comment]
CUDA test failed: https://github.com/opencv/opencv/actions/runs/7224612370/job/19892629635?pr=24704

**Comment by Dhanwanth1803** [comment]
@dkurt Thank you for making the changes. I was not able to work for personal reasons I will continue to contribute soon.

**Pull request merged**